### ISSUE TITLE
Jaws performance and reading alerts

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/john-travoltage#310]
**Test device:**
Dell Laptop
**Operating System:**
Win 10
**Browser:**
Firefox
**Problem description:**
For https://github.com/phetsims/QA/issues/196
When using Jaws and keyboard nav there is a significant drop in performance.  Jaws also very rarely reads the part of the alert when rubbing against the rug that reads the electron count.  After a few minutes it only read twice for me.  Position is read every time.  NVDA does read these alerts consistently.
**Steps to reproduce:**
1. Use Jaws with the sim
2. Use keyboard nav to move the foot.  There will be significant lag
3. After a bit of movement, listen to the alert.  It rarely includes the new electron count.

Troubleshooting information (do not edit):

<details>

Name: ‪John Travoltage‬
URL: https://phet-dev.colorado.edu/html/john-travoltage/1.5.0-rc.1/phet/john-travoltage_en_phet.html
Version: 1.5.0-rc.1 2018-09-25 20:38:20 UTC
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0) Gecko/20100101 Firefox/62.0
Language: en-US
Window: 1536x760
Pixel Ratio: 2.5/1
WebGL: WebGL 1.0
GLSL: WebGL GLSL ES 1.0
Vendor: Mozilla (Mozilla)
Vertex: attribs: 16 varying: 30 uniform: 4096
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 32767x32767
OES_texture_float: true
Dependencies JSON: {"assert":{"sha":"928741cf","branch":"HEAD"},"axon":{"sha":"ea39b709","branch":"HEAD"},"brand":{"sha":"89d28f63","branch":"HEAD"},"chipper":{"sha":"ed169ed3","branch":"HEAD"},"dot":{"sha":"a6af5081","branch":"HEAD"},"john-travoltage":{"sha":"81c3b843","branch":"HEAD"},"joist":{"sha":"79cf90f9","branch":"HEAD"},"kite":{"sha":"a1257832","branch":"HEAD"},"phet-core":{"sha":"2ff7b489","branch":"HEAD"},"phet-io":{"sha":"c1301ac9","branch":"HEAD"},"phet-io-wrapper-classroom-activity":{"sha":"abc4ae3e","branch":"HEAD"},"phet-io-wrapper-hookes-law-energy":{"sha":"429886e7","branch":"HEAD"},"phet-io-wrapper-lab-book":{"sha":"186d07de","branch":"HEAD"},"phet-io-wrappers":{"sha":"9126d1ed","branch":"HEAD"},"phetcommon":{"sha":"33aef589","branch":"HEAD"},"query-string-machine":{"sha":"9ae63aab","branch":"HEAD"},"scenery":{"sha":"eb1234c9","branch":"HEAD"},"scenery-phet":{"sha":"d205340b","branch":"HEAD"},"sherpa":{"sha":"c6a137d3","branch":"HEAD"},"sun":{"sha":"6138c76e","branch":"HEAD"},"tambo":{"sha":"afb95743","branch":"HEAD"},"tandem":{"sha":"1dd1ec01","branch":"HEAD"},"vibe":{"sha":"82f7679d","branch":"HEAD"}}

</details>

**Comment by jessegreenberg** [comment]
I just tried JAWS 2018 on my Windows 10 desktop with FF and everything is very fast. Ill try on my laptop.

**Comment by KatieWoe** [comment]
I did notice that the dropped alerts occurred more often when using keyboard nav.  Enhanced sound seemed to increase their frequency for some reason.  Finally, you can keep adding charges while discharge occurs, and this seems to confuse the screen reader when it tries to say how many charges were lost.

**Comment by jessegreenberg** [comment]
I can confirm on my laptop that JAWS + FF rarely reads the change in charge alert, though sim performance still feels fast. This may be related to a simlar issue in RIAW https://github.com/phetsims/resistance-in-a-wire/issues/165

**Comment by jessegreenberg** [comment]
This wasn't caught in the last release because the previous release supported JAWS + IE11 instead of JAWS + FF.

Since I can't reproduce on my faster desktop, this is a performance related issue like the one in resistance in a wire. JAWS + FF has a hard time reading off the alert immediately after a change in `aria-valuetext`.

**Comment by jessegreenberg** [comment]
I have a few ideas about how to improve this. I think we can generally improve utteranceQueue and make it so we don't have to do a deep dive into sim performance every time this comes up, which has been in RIAW, BASE, and JT since FF Quantum was released.

@emily-phet is it OK if I work on this, and should this be improved for the republication of this sim? This doesn't look complicated, should take an hour or two.

**Comment by emily-phet** [comment]
@jessegreenberg If you think the improvements are generalizable, so will save time in the future, go for it!

**Comment by jessegreenberg** [comment]
Thanks! If I strike out or it takes longer Ill check in again.

**Comment by jessegreenberg** [comment]
Overall, JAWS has trouble with aria-live immediately after announcing an update from aria-valuetext. This was improved by adding a delay to Utterances in the queue to allow the aria-valuetext to finish before announcing the alert.

The above commits improve behavior quite a bit. As @KatieWoe mentioned, JAWS was reading the electron count very rarely before. Now I get the alerts almost every time. I still usually miss the alerts if I rub the foot back and forth, like rapidly pressing "left, left, left, right, right, right, left, left left".

I think the next thing to try would be to try adding the electron count info to the aria-valuetext, rather than going through aria-live at all. @emily-phet is this something we should improve now?

**Comment by emily-phet** [comment]
@jessegreenberg Thanks for the update. I agree that moving electron count into aria-valuetext would be something good to try, though I think we should wait on this until the next iteration of a new feature for JT. Especially with the new sonification, there will be additional auditory feedback that the charges are going onto the body, so hopefully the improvements you've just made plus this sound feedback will result in a good experience most of the time for many users. 

I think moving the electron count into the aria-valuetext is something we'd want to try out with users in interviews, so we should fold that into a future design iteration, rather than doing that right now.

**Comment by jessegreenberg** [comment]
OK, sounds good. So to include these changes we will need a release branch in scenery-phet, plus include https://github.com/phetsims/john-travoltage/commit/ead3cbe9694eb3b371fe966a1e0a4eeff62be9e6 in the 1.5 release branch.

**Comment by KatieWoe** [comment]
Performance is still lagging for me in 1.5.0-rc.2.  The number of electrons reads out more often, but https://github.com/phetsims/john-travoltage/issues/315 has become an issue.  This also means that the count is more likely to be read out if only one "step" has been made. @jessegreenberg

**Comment by jessegreenberg** [comment]
On Slack, @KatieWoe said:
> JAWS I'm still seeing performance issues and it still isn't reading the electron count unless I only move the leg one position and pick up new charges.

So by moving the leg before electron count is read out, JAWS is still not providing the alert to the user because it is trying to read the aria-valuetext. @emily-phet how should we proceed with this issue? In https://github.com/phetsims/john-travoltage/issues/310#issuecomment-425502414 it sounds like further work could wait until the next iteration. Is that still true or should we try other enhancements before publishing?

**Comment by emily-phet** [comment]
So, am I understanding this correctly that essentially one needs to pick up one electron at a time in order to hear the electron count read out? When interacting slowly (one electron at a time), is the readout of alerts otherwise fine? 

Note: For future reference, I read this blog post from Sept. 2018 as I was considering how to respond:
http://blog.freedomscientific.com/firefoxesr/

**Comment by KatieWoe** [comment]
@emily-phet essentially, yes.  Readout is still a bit off, but much better than it was, when moving slowly.

**Comment by jessegreenberg** [comment]
@emily-phet that is correct. It depends on how fast your machine is. On my speedy desktop this is not an issue, but this is the case for my laptop.

On my slower laptop I tested john-travoltage rc.3 with JAWS and Chrome and performance is very smooth and this issue is not present at all.

**Comment by emily-phet** [comment]
Ok - let's go ahead with deploying JT this week, and as a next step (for after deploying) open an issue regarding moving electron count into aria-value text. Then we should discuss this possible change with @terracoda before proceeding further there.

**Comment by terracoda** [comment]
Great news about a re-deployment for this sim.

**Comment by jessegreenberg** [comment]
Sounds good, I made a new issue to discuss and track using aria-valuetext as an enhancement for FF+JAWS. Its on the meeting agenda, but I also think it would be helpful to check in on whether FF is still the best browser to target for use with JAWS.

**Comment by jessegreenberg** [comment]
Closing this issue, we will pick up next steps in https://github.com/phetsims/john-travoltage/issues/318.
